discussions-store/discussions-store-default/src/main/java/org/xwiki/contrib/discussions/store/internal/PageHolderReferenceFactory.java

```diff
@@ -0,0 +1,124 @@
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.StringUtils;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.contrib.discussions.DiscussionStoreConfigurationParameters;
+import org.xwiki.contrib.discussions.domain.references.DiscussionContextEntityReference;
+import org.xwiki.contrib.discussions.domain.references.DiscussionReference;
+import org.xwiki.contrib.discussions.store.DiscussionStoreConfiguration;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.SpaceReference;
+
+/**
+ * Utility component in charge of generating unique references for the various entities to store.
+ *
+ * @version $Id$
+ * @since 2.1
+ */
+@Component(roles = PageHolderReferenceFactory.class)
+@Singleton
+public class PageHolderReferenceFactory
+{
+    /**
+     * Entities supported for reference creation.
+     */
+    enum DiscussionEntity
+    {
+        /**
+         * For storing a {@link org.xwiki.contrib.discussions.domain.DiscussionContext}.
+         */
+        DISCUSSION_CONTEXT,
+
+        /**
+         * For storing a {@link org.xwiki.contrib.discussions.domain.Discussion}.
+         */
+        DISCUSSION,
+
+        /**
+         * For storing a {@link org.xwiki.contrib.discussions.domain.Message}.
+         */
+        MESSAGE
+    }
+
+    @Inject
+    private DiscussionStoreConfigurationFactory discussionStoreConfigurationFactory;
+
+    /**
+     * Generate a unique reference for storing the requested entity at the space location retrieved from the given
+     * parameters.
+     * <p>
+     * The type of the given reference is linked to the requested entity, following this mapping:
+     * <ul>
+     *     <li>DiscussionContext: the reference should be of type {@link DiscussionContextEntityReference}</li>
+     *     <li>Discussion: the reference is not used and can be null</li>
+     *     <li>Message: the reference should be of type {@link DiscussionReference}</li>
+     * </ul>
+     * </p>
+     *
+     * @param entity the entity to store
+     * @param name the name of the entity
+     * @param applicationHint the hint of the application used to retrieve the configuration store
+     * @param reference the linked reference for storing the entity: the type is related to the entity
+     * @param configurationParameters the parameters to use for chosing the storage location
+     * @return a unique reference for storing the entity
+     */
+    public DocumentReference createPageHolderReference(DiscussionEntity entity, String name, String applicationHint,
+        Object reference, DiscussionStoreConfigurationParameters configurationParameters)
+    {
+        String generatedString = UUID.randomUUID().toString();
+        DiscussionStoreConfiguration discussionStoreConfiguration =
+            this.discussionStoreConfigurationFactory.getDiscussionStoreConfiguration(applicationHint);
+
+        SpaceReference spaceLocation;
+        switch (entity) {
```

Comment:
Not a fan of this closed enumeration. One alternative is to let one of the parameter be a lambda (probably a `Callable`), and to simply call it to initialize `spaceLocation`.

Reply:
This is only internal and not supposed to evolve. Also I really don't want to put back retrieving the space location in the different services as it's really similar from one entity to another.

Reply:
Sounds good, it's a net improvement anyway :)